compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2537,4 +2564,160 @@ public int hashCode() {
+            default:
+                return false;
+        }
+    }
+
+    private boolean checkFillerValue(BObjectType type) {
+        if ((type.tsymbol.flags & Flags.ABSTRACT) == Flags.ABSTRACT) {
+            return false;
+        }
+
+        BAttachedFunction initFunction = ((BObjectTypeSymbol) type.tsymbol).initializerFunc;
+        if (initFunction == null) {
+            return true;
+        }
+        if (initFunction.symbol.getReturnType().getKind() != TypeKind.NIL) {
+            return false;
+        }
+
+        for (BVarSymbol bVarSymbol : initFunction.symbol.getParameters()) {
+            if (!bVarSymbol.defaultableParam) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * This will handle two types. Singleton : As singleton can have one value that value should it self be a valid fill
+     * value Union : 1. if nil is a member it is the fill values 2. else all the values should belong to same type and
+     * the default value for that type should be a member of the union precondition : value space should have at least
+     * one element
+     *
+     * @param type BFiniteType union or finite
+     * @return boolean whether type has a valid filler value or not
+     */
+    private boolean checkFillerValue(BFiniteType type) {
+        if (type.isNullable()) {
+            return true;
+        }
+
+        // For singleton types, that value is the implicit initial value
+        if (type.getValueSpace().size() == 1) {
+            return true;
+        }
+
+        boolean defaultFillValuePresent = false;
+
+        Iterator iterator = type.getValueSpace().iterator();
+        BLangExpression firstElement = (BLangExpression) iterator.next();
+        BType firstElementType = firstElement.type;
+        String defaultFillValue = getDefaultFillValue(firstElement);
+        if (firstElement.toString().equals(defaultFillValue)) {
+            defaultFillValuePresent = true;
+        }
+
+        while (iterator.hasNext()) {
+            Object value =  iterator.next();
+            BType valueType = ((BLangExpression) value).type;
+            if (!isSameType(valueType, firstElementType)) {
+                return false;
+            }
+            if (!defaultFillValuePresent && value.toString().equals(defaultFillValue)) {
+                defaultFillValuePresent = true;
+            }
+        }
+        return defaultFillValuePresent;
+    }
+
+    private boolean checkFillerValue(BUnionType type) {
+        if (type.isNullable()) {
+            return true;
+        }
+        Iterator<BType> iterator = type.getMemberTypes().iterator();
+        BType firstMember = iterator.next();
+        boolean defaultFillValuePresent = false;
+
+        // is first value is a valid fill value
+        String defaultFillValue = getDefaultFillValue(firstMember);
+        if (firstMember.toString().equals(defaultFillValue)) {
```

Comment:
Can we validate if this works for constants?

```ballerina
const ZERO = 0;

type Foo ZERO|1|2;
```

Reply:
fixed test case added 
https://github.com/ballerina-platform/ballerina-lang/pull/20749/commits/7d525fc1a7d35deb2cffeed2a9d4e3c5022d5b8b